### PR TITLE
[feat]: [CI-12621]: Parse Buildx Logs to get cache metrics

### PR DESCRIPTION
**Pull request by anurag-harness**
We need to parse the output of the docker buildx command and calculate cache metrics using that. In this PR we use a multipart writer to direct the output of the buildx command to stdout and a pipe. We read from the pipe and parse the output to get cache metrics which we save to a file. 
If the file path is not given we execute the command as before

Here is a screenshot of a successful execution

Screenshot 2024-05-23 at 4 48 42 PM

### CODE REVIEW

File: docker.go [hunk 1]
```diff
@@ -281,6 +330,63 @@ func (p Plugin) Exec() error {
...
+
+	dir := filepath.Dir(filename)
+	err = os.MkdirAll(dir, 0644)
+	if err != nil {
+		return fmt.Errorf("failed with err %s to create %s directory for artifact file", err, dir)
+	}
+
+	err = os.WriteFile(filename, b, 0644)
+	if err != nil {
+		return fmt.Errorf("failed to write artifact to artifact file %s", filename)
```
Comment (rutvijmehta-harness) [inline docker.go:385]: Change the error string everywhere

File: docker.go [hunk 2]
```diff
@@ -232,8 +244,45 @@ func (p Plugin) Exec() error {
...
+				return err
+			}
 
-		err := cmd.Run()
+			// Handle errors from the command goroutine
+			if err := <-errChan; err != nil {
+				return err
+			}
+
+			if err := saveCacheMetrics(cacheMetrics, p.CacheMetricsFile); err != nil {
```
Comment (rutvijmehta-harness) [inline docker.go:280]:
```suggestion
			if err := writeCacheMetrics(cacheMetrics, p.CacheMetricsFile); err != nil {
```

File: docker.go [hunk 3]
```diff
@@ -281,6 +330,63 @@ func (p Plugin) Exec() error {
...
 }
 
+func parseCacheMetrics(r io.Reader) (CacheMetrics, error) {
+	scanner := bufio.NewScanner(r)
+	var cacheMetrics CacheMetrics
+
+	// Regular expression to match the log lines
+	re := regexp.MustCompile(`#\d+ (DONE|CACHED|ERRORED|CANCELLED)`)
+
+	for scanner.Scan() {
```
[the remark this answers is not in the document]
Comment (anurag-harness) [reply docker.go:340]: The reader is blocked until the pipe writer is closed. Once the pipe writer in closed in our separate goroutine the reader will get EOF and exit: https://pkg.go.dev/io#PipeReader.Read. Hence Synchronization is handled automatically with the pipe here

File: docker.go [hunk 4]
```diff
@@ -232,8 +244,45 @@ func (p Plugin) Exec() error {
 		cmd.Stdout = os.Stdout
 		cmd.Stderr = os.Stderr
 		trace(cmd)
+		var err error
+		if isCommandBuildxBuild(cmd.Args) && p.CacheMetricsFile != "" {
+			errChan := make(chan error, 1)
```
[the remark this answers is not in the document]
Comment (anurag-harness) [reply docker.go:249]: Yes makes sense. We dont need a wait as well since the pipe will handle the synchronization here

File: docker.go (hunk 1 above)
Comment (anurag-harness) [reply docker.go]: Done

File: docker.go (hunk 2 above)
Comment (anurag-harness) [reply docker.go:276]: Done

**Comment by anurag-harness** [comment]
New PR: https://github.com/drone-plugins/drone-buildx/pull/22

**Pull request closed without merging**